File: spec/account_spec.rb
require 'account'

describe Account do
  
  let(:account) { Account.new([]) }
  let(:entry) { double(:entry) }

  it { is_expected.to respond_to(:entries) }

  it 'is initialized with an empty array of entries' do
    expect(account.entries).to match_array([])
  end

  describe '#add_entry' do
    it 'adds an entry to the entries array' do
      expect{ account.add_entry(entry) }.to change{ account.entries.length }.by(1)
    end
  end

  describe '#get_entry_index' do
    it "retrieves the index of an entry in the entries array" do
      account.add_entry(entry)
      p account
      expect(account.get_entry_index(entry)).to eq 0
    end
  end
end

File: spec/accountSpec.js
'use strict';

var Account = require('../src/account').Account
var validateDate = require('../src/account').validateDate
var convertStringToDate = require('../src/transaction').convertStringToDate

describe('Account', function () {
  var account, date, transaction, dateConverted;
  beforeEach(function () {
    date = '10/01/2012'
    dateConverted = convertStringToDate(date)
    account = new Account();
  });

  describe('is initialized', function () {
    it('with an empty array of transactions', function () {
      expect(account.transactions).not.toBe(null);
    });
  });

  describe('#recordTransaction', function () {
    it('records a new transaction in the account', function () {
      account.recordTransaction(100, date);
      transaction = account.transactions.pop();
      expect(transaction.date.getTime()).toBe(dateConverted.getTime());
      expect(transaction.amount).toBe(100);
    });
  });

  describe('#madeDeposit', function () {
    it('can add a credit transaction', function () {
      var length = account.transactions.length;
      account.madeDeposit(100, date);
      expect(account.transactions.length).toBe(length + 1);
    });
    it('throws an error for invalid dates', function () {
      expect(function () { account.madeDeposit(100, '1/1/2012') }).toThrow('Please enter a date in dd/mm/yyyy format');
    });
  });

  describe('#withdrewMoney', function () {
    it('can add a debit transaction', function () {
      var length = account.transactions.length;
      account.withdrewMoney(100, date);
      expect(account.transactions.length).toBe(length + 1);
    });
    it('throws an error for invalid dates', function () {
      expect(function () { account.madeDeposit(100, '1/14/2012') }).toThrow('Please enter a date in dd/mm/yyyy format');
    });
  });
})
